lib/elixir/lib/map_set.ex

```diff
@@ -226,7 +231,7 @@ defmodule MapSet do
       #MapSet<[]>
 
   """
-  @spec intersection(t, t) :: t
+  @spec intersection(t(val1), t(val2)) :: t(val1 | val2) when val1: value, val2: value
```

Comment:
This type spec is quite confusing because only overlapping types can be returned. If `val1` and `val2` are disjoint sets then the intersection is the empty set. Given we can't describe this, it would be clearer to have `intersection(t(val), t(val)) :: t(val) when val: value`.